app/models/cherrypick_task.rb

```diff
@@ -91,8 +91,72 @@ def full?
       @wells.size == @size
     end
 
+    # Creates control requests for the control assets provided and adds them to the batch
+    def create_control_requests!(batch, control_assets)
+      control_requests = control_assets.map do |control_asset|
+        CherrypickRequest.create(
+          asset: control_asset,
+          target_asset: Well.new,
+          submission: batch.requests.first.submission,
+          request_type: batch.requests.first.request_type,
+          request_purpose: 'standard'
+        )
+      end
+      batch.requests << control_requests
+      control_requests
+    end
+
+    # Creates a new control request for the control_asset and adds it into the current_destination_plate plate
+    def add_control_request(batch, control_asset)
+      control_request = create_control_requests!(batch, [control_asset]).first
+      control_request.start!
+      push(control_request.id, control_request.asset.plate.human_barcode, control_request.asset.map_description)
+    end
+
+    # Adds any consecutive list of control requests into the current_destination_plate
+    def add_any_consecutive_control_requests(control_positions, batch, control_assets)
+      # find the index of the well we are filling right now
+      current_well_index = content.length
+
+      # check if this well should contain a control
+      # add it if so, and any consecutive ones by looping
+      while control_positions.include?(current_well_index)
+        control_asset = control_assets[control_positions.find_index(current_well_index)]
+        add_control_request(batch, control_asset)
+        # above adds to current_destination_plate, so current_well_index should be recalculated
+        current_well_index = content.length
+      end
+    end
+
+    # Adds any remaining control requests not already added, into the current_destination_plate plate
+    def add_remaining_control_requests(control_positions, batch, control_assets)
+      control_positions.each_with_index do |pos, idx|
+        if pos >= content.length
+          control_asset = control_assets[idx]
+          add_control_request(batch, control_asset)
+        end
+      end
+    end
+
     def push(request_id, plate_barcode, well_location)
       @wells << [request_id, plate_barcode, well_location]
+
+      add_any_wells_from_template_or_partial(@wells)
+      self
+    end
+
+    def remaining_wells(control_positions)
+      remaining_controls = control_positions.select { |c| c > @wells.length }
+      remaining_used_wells = @used_wells.keys.select { |c| c > @wells.length }
+      remaining_controls.concat(remaining_used_wells).flatten
+    end
+
+    def push_with_controls(request_id, plate_barcode, well_location, control_positions, batch, control_assets)
+      @wells << [request_id, plate_barcode, well_location]
+      if control_positions
```

Comment:
We think `if control_positions` will always return true, because an empty list is truthy

Reply:
add comment `# would be nil if no control plate selected`

Reply:
Added